##### ARTI 560 - Computer Vision

## Image Classification with Vision Transformer (ViT) - Exercise

### Objective

In this exercise, you will test the pretrained Vision Transformer (ViT) model on 5 real-world images that you find online.

You will:

1. Download 5 images for different classes in [ImageNet](https://github.com/Waikato/wekaDeeplearning4j/blob/master/docs/user-guide/class-maps/IMAGENET.md).

2. Load the ImageNet class names from a [text file](https://raw.githubusercontent.com/pytorch/hub/master/imagenet_classes.txt).

3. Use ViT to predict the class for each image.

4. Record whether the prediction was correct.

#### Important Note

For this exercise, you MUST use the following KerasHub components:

- [keras_hub.models.ViTImageClassifier](https://keras.io/keras_hub/api/models/vit/vit_image_classifier/)

- [keras_hub.models.ViTImageClassifierPreprocessor](https://keras.io/keras_hub/api/models/vit/vit_image_classifier_preprocessor/)

This ensures your input preprocessing (resizing + normalization) matches what the pretrained ViT model expects.

Do not replace the preprocessor with manual normalization (such as dividing by 255), because it may produce incorrect predictions.

In [6]:
import requests
import numpy as np
from PIL import Image
import keras
import keras_hub

# Load preprocessor
preprocessor = keras_hub.models.ViTImageClassifierPreprocessor.from_preset(
    "vit_base_patch16_224_imagenet"
)

# Load model
model = keras_hub.models.ViTImageClassifier.from_preset(
    "vit_base_patch16_224_imagenet",
    preprocessor=preprocessor
)

# Load ImageNet class names
classes_url = "https://raw.githubusercontent.com/pytorch/hub/master/imagenet_classes.txt"
classes = requests.get(classes_url).text.splitlines()

# Define images (تأكدي الصور موجودة في نفس المسار)
TEST_IMAGES = [
    {"path": "/content/elephant.jpg", "true_class": "African elephant"},
    {"path": "/content/Espresso.webp", "true_class": "espresso"},
    {"path": "/content/volcano.webp", "true_class": "volcano"},
    {"path": "/content/tennis_ball.png", "true_class": "tennis ball"},
    {"path": "/content/Cheeseburger.jpg", "true_class": "cheeseburger"},
]

print("===== ViT Predictions =====")

for i, item in enumerate(TEST_IMAGES, start=1):
    print(f"\n-- Image {i} --")

    img = Image.open(item["path"]).convert("RGB")
    img_array = np.array(img, dtype=np.uint8)
    img_batch = np.expand_dims(img_array, axis=0)

    logits = model.predict(img_batch, verbose=0)
    probs = keras.ops.softmax(logits, axis=-1)
    probs_np = np.array(probs)[0]

    predicted_id = int(np.argmax(probs_np))
    predicted_label = classes[predicted_id]
    confidence = float(probs_np[predicted_id])

    correct = "Yes" if item["true_class"].lower() in predicted_label.lower() else "No"

    print(f"Image File : {item['path']}")
    print(f"True Label : {item['true_class']}")
    print(f"Predicted  : {predicted_label} (conf: {confidence:.2%})")
    print(f"Correct?   : {correct}")

===== ViT Predictions =====

-- Image 1 --
Image File : /content/elephant.jpg
True Label : African elephant
Predicted  : African elephant (conf: 83.16%)
Correct?   : Yes

-- Image 2 --
Image File : /content/Espresso.webp
True Label : espresso
Predicted  : espresso (conf: 99.15%)
Correct?   : Yes

-- Image 3 --
Image File : /content/volcano.webp
True Label : volcano
Predicted  : volcano (conf: 99.97%)
Correct?   : Yes

-- Image 4 --
Image File : /content/tennis_ball.png
True Label : tennis ball
Predicted  : tennis ball (conf: 99.95%)
Correct?   : Yes

-- Image 5 --
Image File : /content/Cheeseburger.jpg
True Label : cheeseburger
Predicted  : cheeseburger (conf: 99.78%)
Correct?   : Yes


### Record Your Results

Fill the table below based on your results:
| Image File        | Predicted Label              | True Label (What you searched) | Correct? (Yes/No) |
|-------------------|------------------------------|---------------------------------|-------------------|
| elephant.jpg      | African elephant (83.16%)    | African elephant                | Yes               |
| Espresso.webp     | espresso (99.15%)            | espresso                        | Yes               |
| volcano.webp      | volcano (99.97%)             | volcano                         | Yes               |
| tennis_ball.png   | tennis ball (99.95%)         | tennis ball                     | Yes               |
| Cheeseburger.jpg  | cheeseburger (99.78%)        | cheeseburger                    | Yes               |